x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java

```diff
@@ -408,34 +399,66 @@ private int getRegionSize(long fileLength, int region) {
         return effectiveRegionSize;
     }
 
-    Entry<CacheFileRegion> get(KeyType cacheKey, long fileLength, int region) {
+    CacheFileRegion get(KeyType cacheKey, long fileLength, int region) {
         final RegionKey<KeyType> regionKey = new RegionKey<>(cacheKey, region);
-        final long now = threadPool.relativeTimeInMillis();
         // try to just get from the map on the fast-path to save instantiating the capturing lambda needed on the slow path if we did not
         // find an entry
-        var entry = keyMapping.get(regionKey);
+        var entry = cache.get(regionKey);
         if (entry == null) {
             final int effectiveRegionSize = getRegionSize(fileLength, region);
-            entry = keyMapping.computeIfAbsent(regionKey, key -> new Entry<>(new CacheFileRegion(key, effectiveRegionSize), now));
+            entry = cache.computeIfAbsent(regionKey, key -> new CacheFileRegion(key, effectiveRegionSize));
         }
         // io is volatile, double locking is fine, as long as we assign it last.
-        if (entry.chunk.io == null) {
-            synchronized (entry.chunk) {
-                if (entry.chunk.io == null) {
+        if (entry.io == null) {
+            synchronized (entry) {
+                if (entry.io == null) {
                     return initChunk(entry);
                 }
             }
         }
-        assert assertChunkActiveOrEvicted(entry);
+        return entry;
+    }
 
-        // existing item, check if we need to promote item
-        if (now - entry.lastAccessed >= minTimeDelta) {
-            maybePromote(now, entry);
+    private CacheFileRegion initChunk(CacheFileRegion entry) {
+        assert Thread.holdsLock(entry);
+        RegionKey<KeyType> regionKey = entry.regionKey;
+        if (cache.get(regionKey) != entry) {
+            throwAlreadyClosed("no free region found (contender)");
+        }
+        final SharedBytes.IO freeSlot = freeRegions.poll();
+        if (freeSlot != null) {
+            // no need to evict an item, just add
+            assignToSlot(entry, freeSlot);
+        } else {
+            // need to evict something
+            cache.evict();
+            final SharedBytes.IO freeSlotRetry = freeRegions.poll();
+            if (freeSlotRetry != null) {
+                assignToSlot(entry, freeSlotRetry);
+            } else {
+                synchronized (this) {
+                    cache.evictMatching(r -> r == regionKey);
```

Comment:
This is a potentially massive slowdown relative to the previous version doing a map remove instead isn't it? We'd have to iterate all the pages to do this? (which would be uncool if we wanted to have more of those and smaller ones in the future :))